File: Server/Models/Entities/Role.cs
using System.ComponentModel.DataAnnotations;

namespace Authentication.Models.Entities;

public class Role
{
    [Key] public int RoleId { get; set; }

    public string RoleName { get; set; }
    public int RoleType { get; set; }
    public required string Description { get; init; }
    public Guid? CreatedBy { get; set; }

    public RoleDto ToDto()
    {
        return new RoleDto
        {
            RoleId = RoleId,
            RoleName = RoleName,
            CreatedBy = CreatedBy,
            Description = Description
        };
    }
}

public class RoleDto
{
    public int RoleId { get; set; }
    public required string RoleName { get; init; }
    public required string Description { get; init; }
    public Guid? CreatedBy { get; set; }
    public List<int> Scopes { get; set; }
}

public class CreateRoleDto
{
    public string Name { get; set; }
    public string Slug { get; set; }
    public int RoleType { get; set; }
    public string Description { get; init; }
    public Guid CreatedBy { get; set; }
}

public class UpdateRoleDto
{
    public string RoleName { get; set; }
    public string Slug { get; set; }
    public string Description { get; set; }
    public int RoleType { get; set; }
    public Guid CreatedBy { get; set; }
}

File: website/models/role.ts
import {z} from "zod";

// Matches the backend RoleDto (Server/Models/Entities/Role.cs)
export const RoleSchema = z.object({
    roleId: z.number(),
    roleName: z.string(),
    description: z.string(),
    createdBy: z.string().uuid().nullable().optional(),
    scopes: z.array(z.number()).optional(),
});

export type Role = z.infer<typeof RoleSchema>;

// Matches the backend CreateRoleDto — roleType and slug are derived server-side
export const CreateRoleSchema = z.object({
    name: z.string(),
    description: z.string().optional(),
    slug: z.string().optional(),
});

export type CreateRoleDto = z.infer<typeof CreateRoleSchema>;

export const UpdateRoleSchema = z.object({
    roleName: z.string(),
    slug: z.string(),
    description: z.string(),
    roleType: z.number(),
});

export type UpdateRoleDto = z.infer<typeof UpdateRoleSchema>;

// Matches UpdateRoleScopesDto — PATCH /teams/{teamId}/roles/{roleId}/permissions
export type UpdateRoleScopesDto = {
    scopes: number[];
};
